src/script/sign.h

```diff
@@ -67,9 +67,349 @@ struct SignatureData {
     explicit SignatureData(const CScript& script) : scriptSig(script) {}
 };
 
+// Note: These constants are in reverse byte order because serialization uses LSB
+static const uint32_t PSBT_MAGIC_BYTES = 0x74627370;
+static const uint8_t PSBT_UNSIGNED_TX = 0x00;
+static const uint8_t PSBT_REDEEMSCRIPT = 0x01;
+static const uint8_t PSBT_WITNESSSCRIPT = 0x02;
+static const uint8_t PSBT_NON_WITNESS_UTXO = 0x03;
+static const uint8_t PSBT_WITNESS_UTXO = 0x04;
+static const uint8_t PSBT_PARTIAL_SIG = 0x05;
+static const uint8_t PSBT_BIP32_KEYPATH = 0x06;
+
+// The separator is 0x00. Reading this in means that the unserializer can interpret it
+// as a 0 length key. which indicates that this is the separator. The separator has no value.
+static const uint8_t PSBT_SEPARATOR = 0x00;
+
+/** An structure for PSBTs which contain per input information */
+struct PartiallySignedInput
+{
+    CTransactionRef non_witness_utxo;
+    CTxOut witness_utxo;
+    std::map<CPubKey, std::vector<unsigned char>> partial_sigs;
+    std::map<uint32_t, std::vector<uint32_t>> hd_keypaths;
+    std::map<std::vector<unsigned char>, std::vector<unsigned char>> unknown;
+
+    PartiallySignedInput(): non_witness_utxo(), witness_utxo(), partial_sigs(), hd_keypaths(), unknown() {}
+    void SetNull();
+    bool IsNull();
+};
+
+/** A version of CTransaction with the PSBT format*/
```

Comment:
```c++
You can add something like 

template<typename X>
std::vector<unsigned char> SerializeToVector(uint8_t header, const X& x)
{
    std::vector<unsigned char> ret;
    CVectorWriter ss(SER_DISK, CLIENT_VERSION, ret, 0);
    ss << header << x;
    return ret;
}
```

Then you can use `s << SerializeToVector(PSBT_REDEEMSCRIPT, entry.first);` instead of
```c++
WriteCompactSize(s, entry.first.size() + sizeof(PSBT_REDEEMSCRIPT));
s << PSBT_REDEEMSCRIPT;
s << entry.first;
```
and many other similar cases.